src/Npgsql/ClusterAwareDataSource.cs

```diff
@@ -0,0 +1,736 @@
+//                             Debug.Assert(databaseState != DatabaseState.Unknown);
+//                             if (!stateValidator(databaseState, preferredType))
+//                             {
+//                                 pool.Return(connector);
+//                                 continue;
+//                             }
+//                         }
+//
+//                         return connector;
+//                     }
+//                 }
+//             }
+//             catch (Exception ex)
+//             {
+//                 exceptions.Add(ex);
+//                 if (connector is not null)
+//                     pool.Return(connector);
+//             }
+//         }
+//
+//         return null;
+//     }
+//     
+//     async ValueTask<NpgsqlConnector?> TryGet(
+//         NpgsqlConnection conn,
+//         TimeSpan timeoutPerHost,
+//         bool async,
+//         TargetSessionAttributes preferredType,
+//         Func<DatabaseState, TargetSessionAttributes, bool> stateValidator,
+//         int poolIndex,
+//         IList<Exception> exceptions,
+//         CancellationToken cancellationToken)
+//     {
+//         var pools = _pools;
+//         for (var i = 0; i < pools.Length; i++)
+//         {
+//             var pool = pools[poolIndex];
+//             poolIndex++;
+//             if (poolIndex == pools.Length)
+//                 poolIndex = 0;
+//
+//             var databaseState = pool.GetDatabaseState();
+//             if (!stateValidator(databaseState, preferredType))
+//                 continue;
+//
+//             NpgsqlConnector? connector = null;
+//
+//             try
+//             {
+//                 connector = await pool.Get(conn, new NpgsqlTimeout(timeoutPerHost), async, cancellationToken);
+//                 if (databaseState == DatabaseState.Unknown)
+//                 {
+//                     // Get might have opened a new physical connection and refreshed the database state, check again
+//                     databaseState = pool.GetDatabaseState();
+//                     if (databaseState == DatabaseState.Unknown)
+//                         databaseState = await connector.QueryDatabaseState(new NpgsqlTimeout(timeoutPerHost), async, cancellationToken);
+//
+//                     Debug.Assert(databaseState != DatabaseState.Unknown);
+//                     if (!stateValidator(databaseState, preferredType))
+//                     {
+//                         pool.Return(connector);
+//                         continue;
+//                     }
+//                 }
+//
+//                 return connector;
+//             }
+//             catch (Exception ex)
+//             {
+//                 exceptions.Add(ex);
+//                 if (connector is not null)
+//                     pool.Return(connector);
+//             }
+//         }
+//
+//         return null;
+//     }
+//
+// }
+
+
+/*
+ * ------------------------------
+ * Approach 2
```

Comment:
Can you please explain what the two approaches are briefly and why you are going with #2? You can put it in PR summary itself.